src/main/scala/is/hail/utils/Py4jUtils.scala

```diff
@@ -49,4 +55,56 @@ trait Py4jUtils {
   def makeDouble(f: Float): Double = f.toDouble
 
   def makeDouble(d: Double): Double = d
+
+  def readFile(path: String, hc: HailContext): HadoopPyReader = hc.hadoopConf.readFile(path) { in =>
+    new HadoopPyReader(hc.hadoopConf.unsafeReader(path))
+  }
+
+  def writeFile(path: String, hc: HailContext): HadoopPyWriter = {
+    new HadoopPyWriter(hc.hadoopConf.unsafeWriter(path))
+  }
+
+  def copyFile(from: String, to: String, hc: HailContext) {
+    hc.hadoopConf.copy(from, to)
+  }
 }
+
+class HadoopPyReader(in: InputStream) {
+  var eof = false
+
+  def read(n: Int): String = {
+    if (eof)
+      new String(new Array[Byte](0), "ISO-8859-1")
+    else {
+      var i = 0
+      val b = new Array[Byte](n)
+      while (!eof && i < n) {
```

Comment:
I don't think this is the intended behavior.  You should just call in.read() once.  From https://docs.python.org/2/library/io.html#io.RawIOBase:

 > read(n=-1)
 > Read up to n bytes from the object and return them. As a convenience, if n is unspecified or -1, readall() is called. Otherwise, only one system call is ever made. Fewer than n bytes may be returned if the operating system call returns fewer than n bytes.

and read is implemented in terms of readinto:

https://github.com/python/cpython/blob/1e73dbbc29c96d0739ffef92db36f63aa1aa30da/Lib/_pyio.py#L569

So readinto should make one "system call".

This should simplify the code a bit.